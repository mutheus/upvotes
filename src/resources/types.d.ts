declare module 'feeds' {
  export type FeedType = {
    id: number,
    author: {
      username: string
    }
    content: string
    likes: number
    loves: number
    createdAt: string
  }
}
