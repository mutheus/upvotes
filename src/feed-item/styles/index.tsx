import styled from 'styled-components/macro'

export const FeedItemContainer = styled.div`
  display: flex;
  padding: 0 2em 2em;
  gap: 1em;

  img {
    width: 32px;
    height: 32px;
  }

  > div {
    display: flex;
    flex-direction: column;
    width: 100%;

    p {
      word-break: break-all;
    }
  }
`

export const UserAvatar = styled.img`
  width: 32px;
  height: 32px;
  border-radius: 50%;
  object-fit: cover;
`

export const UserName = styled.p`
  margin: .3em 0;
  color: ${({ theme }) => theme.colors.darkGray};
  font-size: .95rem;
  font-weight: 600;
`

export const CreatedAt = styled.p`
  color: ${({ theme }) => theme.colors.darkGray};
  font-size: .95rem;
`

export const Interactions = styled.div`
  font-size: .9rem;
  display: flex;
  gap: 2em;
  font-weight: 500;

  > div {
    display: flex;
    align-items: center;
    gap: 8px;
  }
`

export const LikeBtn = styled.button`
  border: none;
  background-color: transparent;
  padding: 0;
  margin: 0;

  path {
    fill: ${({ theme }) => theme.colors.text};
    stroke: ${({ theme }) => theme.colors.text};
    stroke-width: .6px;
  }
`

export const LoveBtn = styled(LikeBtn)`
  path {
    stroke-width: 0;
  }
`
