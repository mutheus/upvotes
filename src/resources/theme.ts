import { DefaultTheme } from 'styled-components'

export const theme: DefaultTheme = {
  colors: {
    purple: '#8652FF',
    lightPurple: '#9879D2',
    text: '#3E3E3E',
    gray: '#C7C7C7',
    darkGray: '#8a8a8a',
  },
}
