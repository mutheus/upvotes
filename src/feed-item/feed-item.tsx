import { ReactComponent as HeartIcon } from './assets/heart-icon.svg'
import { ReactComponent as LikeIcon } from './assets/like-icon.svg'
import { ReactComponent as HeartIconFilled } from './assets/heart-icon-filled.svg'
import { ReactComponent as LikeIconFilled } from './assets/like-icon-filled.svg'
import BlankUser from 'shared/assets/blank-user.png'
import { FeedType } from 'feeds'
import * as S from './styles'
import { api } from 'services/api'

type FeedItemProps = {
  feed: FeedType
  onInteraction: () => void
}

export function FeedItem ({ feed, onInteraction }: FeedItemProps) {
  const isLikedTrue = !!feed.activeUserLikedIt
  const isLovedTrue = !!feed.activeUserLovedIt

  const handleLikeClick = async (id: number) => {
    await api.post('/reaction', {
      feedId: id,
      like: !isLikedTrue,
    })

    onInteraction()
  }

  const handleLoveClick = async (id: number) => {
    await api.post('/reaction', {
      feedId: id,
      love: !isLovedTrue,
    })

    onInteraction()
  }

  return (
    <S.FeedItemContainer>
      <S.UserAvatar src={BlankUser} alt='Blank user' />

      <S.UserInfo>
        <S.UserName>@{feed.author.username.replace(/[@.]/g, '')}</S.UserName>

        <S.Content>{feed.content}</S.Content>

        <S.CreatedAt>{feed.createdAt}</S.CreatedAt>

        <S.Interactions>
          <div>
            <S.LikeBtn onClick={() => handleLikeClick(feed.id)} isLikedTrue={isLikedTrue}>
              {
                isLikedTrue
                  ? (
                    <LikeIconFilled />
                    )
                  : (
                    <LikeIcon />
                    )
              }
            </S.LikeBtn>

            <span>
              {feed.likes > 1 ? `${feed.likes} Likes` : `${feed.likes} Like`}
            </span>
          </div>

          <div>
            <S.LoveBtn onClick={() => handleLoveClick(feed.id)} isLovedTrue={isLovedTrue}>
              {
                isLovedTrue
                  ? (
                    <HeartIconFilled />
                    )
                  : (
                    <HeartIcon />
                    )
              }
            </S.LoveBtn>

            <span>
              {feed.loves > 1 ? `${feed.loves} Loves` : `${feed.loves} Love`}
            </span>
          </div>
        </S.Interactions>
      </S.UserInfo>
    </S.FeedItemContainer>
  )
}
