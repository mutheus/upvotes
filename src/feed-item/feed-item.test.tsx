import { render } from '@testing-library/react'
import { FeedItem } from './feed-item'
import { ThemeProvider } from 'styled-components'
import { theme } from 'resources/theme'

describe('Feed item component', () => {
  const feed = {
    id: 123,
    author: {
      username: 'jõao',
    },
    content: 'A test',
    likes: 2,
    loves: 1,
    createdAt: '2021-11-12T18:22:12.720Z',
    activeUserLikedIt: 0,
    activeUserLovedIt: 1,
  }

  it('The component´s snapshot should exactly match', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <FeedItem
          feed={feed}
          onInteraction={() => {}}
        />
      </ThemeProvider>)

    expect(container.firstChild).toMatchSnapshot()
  })
})
