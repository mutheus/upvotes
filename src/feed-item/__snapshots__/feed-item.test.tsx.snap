// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed item component The component´s snapshot should exactly match 1`] = `
<div
  class="sc-bdvvtL iJtKIV"
>
  <img
    alt="Blank user"
    class="sc-gsDKAQ iirgEA"
    src="blank-user.png"
  />
  <div
    class="sc-dkPtRN hUnxIf"
  >
    <p
      class="sc-eCImPb cZwBKB"
    >
      @
      jõao
    </p>
    <p
      class="sc-hKwDye MUMnp"
    >
      A test
    </p>
    <p
      class="sc-jRQBWg kXWXTQ"
    >
      2021-11-12T18:22:12.720Z
    </p>
    <div
      class="sc-gKclnd ljuhmM"
    >
      <div>
        <button
          class="sc-iCfMLu sc-pVTFL hLUYxi dVdvxL"
        >
          <svg>
            like-icon.svg
          </svg>
        </button>
        <span>
          2 Likes
        </span>
      </div>
      <div>
        <button
          class="sc-iCfMLu sc-furwcr hLUYxi hEBSVL"
        >
          <svg>
            heart-icon-filled.svg
          </svg>
        </button>
        <span>
          1 Love
        </span>
      </div>
    </div>
  </div>
</div>
`;
